statsmodels/gam/gam_penalties.py

```diff
@@ -0,0 +1,309 @@
+            penalty weights
+
+        Returns
+        -------
+        penalty matrix
+            square penalty matrix for quadratic penalization. The number
+            of rows and columns are equal to the number of columns in the
+            smooth terms, i.e. the number of parameters for this smooth
+            term in the regression model
+        """
+        if alpha is None:
+            alpha = self.alpha
+
+        return alpha * self.univariate_smoother.cov_der2
+
+
+class MultivariateGamPenalty(Penalty):
+    """
+    Penalty for Generalized Additive Models
+
+    Parameters
+    -----------
+    multivariate_smoother : instance
+        instance of additive smoother or spline class
+    alpha : list of float
+        default penalty weight, list with length equal to the number of smooth
+        terms. ``alpha`` can also be provided to each method.
+    weights: array-like
+        currently not used
+        is a list of doubles of the same length as alpha or a list
+        of ndarrays where each component has the length equal to the number
+        of columns in that component
+    start_idx : int
+        number of parameters that come before the smooth terms. If the model
+        has a linear component, then the parameters for the smooth components
+        start at ``start_index``.
+
+    Attributes
+    -----------
+    Parameters are stored, additionally
+    nob s: The number of samples used during the estimation
+
+    dim_basis : number of columns of additive smoother. Number of columns
+        in all smoothers.
+    k_variables : number of smooth terms
+    k_params : total number of parameters in the regression model
+
+    """
+
+    def __init__(self, multivariate_smoother, alpha, weights=None,
+                 start_idx=0):
+
+        if len(multivariate_smoother.smoothers) != len(alpha):
+            raise ValueError('all the input values should be list of the same'
+                             ' length. len(smoothers)=',
+                             len(multivariate_smoother.smoothers),
+                             ' len(alphas)=', len(alpha))
+
+        self.multivariate_smoother = multivariate_smoother
+        self.dim_basis = self.multivariate_smoother.dim_basis
+        self.k_variables = self.multivariate_smoother.k_variables
+        self.nobs = self.multivariate_smoother.nobs
+        self.alpha = alpha
+        self.start_idx = start_idx
+        self.k_params = start_idx + self.dim_basis
+
+        # TODO: Review this,
+        if weights is None:
+            # weights should hanve total length as params
+            # but it can also be scalar in individual
+            self.weights = [1. for _ in range(self.k_variables)]
+        else:
+            import warnings
+            warnings.warn('weights is currently ignored')
+            self.weights = weights
+
+        self.mask = [np.array([False] * self.k_params)
+                     for _ in range(self.k_variables)]
```

Comment:
np.zeros, also could just make this a single 2D array?

Reply:
the top level of those are the different spline components, we keep those as a list (currently). They are always used in loops, so list is more efficient, than arrays, and avoids the issue of column versus row orientation.

Reply:
change to list of `zeros`